File: src/LaneKeeping/lanekeeping.py
import configparser
import math
import traceback

import cv2
import numpy as np
from scipy.stats import norm


class LaneKeeping:
    """
    This class implements the lane keeping algorithm by calculating steering angles from the detected lines.

    Attributes:
        log (logging object):
            Logic logger, child of root logger
        config (configparser object):
            Open configuration file of the whole project
        width (integer):
            The width of the input image
        height (integer):
            The height of the input image
        angle (float):
            Steering angle
        last_angle (float):
            Previous steering angle
        steer_value_list (list):
            Store last k steering values
        median_constant (integer):
            Num of the last steering angles to be saved
        print_desire_lane (bool):
            Visualization of the desired lane
        bottom_width (integer):
            Bottom width difference of left and right lane div 2
        top_width (integer)
            Top width difference of left and right lane div 2
        slices (integer):
            Number of slices from the lane detection
        bottom_row_index (integer):
            Height of the first histogram
        top_row_index (integer) :
            Height of the last histogram
    """

    def __init__(self, width, height, logger, var_handler):

        self.width = width
        self.height = height
        self.log = logger
        self.var = var_handler

        self.angle = 0.0
        self.last_angle = 0.0

        self.config = configparser.ConfigParser()
        self.config.read("config.ini")
        self.speedrun = self.config["SPEED"].getboolean("speedrun")

        # Rolling average parameters
        self.steer_value_list = list()
        self.median_constant = int(self.config["LANE_KEEPING"].get("median_constant"))

        self.print_desire_lane = self.config["LANE_KEEPING"].getboolean("print_desire_lane")

        self.max_coef_of_sharp_turn = float(self.config["LANE_KEEPING"].get("max_coef_of_sharp_turn"))
        self.min_coef_of_sharp_turn = float(self.config["LANE_KEEPING"].get("min_coef_of_sharp_turn"))
        self.sharp_turning_factor = float(self.config["LANE_KEEPING"].get("sharp_turning_factor"))
        self.max_lk_steer = int(self.config["PARAMS"].get("max_lk_steer"))

        self.slices = int(self.config["LANE_DETECT"].get("slices"))
        self.bottom_offset = int(self.config["LANE_DETECT"].get("bottom_offset"))

        if self.var.get_lanes_det_cam() == "455":
            self.choose_455()
        elif self.var.get_lanes_det_cam() == "405":
            self.choose_405()

        self.prev_mean_right = 0
        self.prev_mean_left = 0

        self.turn_left = False
        self.turn_right = False

        self.change_lane_stage_1 = True
        self.change_lane_stage_2 = False
        self.count = 0

        self.get_error_weights = range(0, self.real_slices)

        mu = float(self.config["LANE_KEEPING"].get("mu"))
        sigma = float(self.config["LANE_KEEPING"].get("sigma"))

        self.max_smoothing_iterations = int(self.config["CHANGE_LANE"].get("max_smoothing_iterations"))
        self.change_lane_threshold = int(self.config["CHANGE_LANE"].get("threshold"))
        self.change_lane_count = int(self.config["CHANGE_LANE"].get("count"))
        self.change_lane_mid = {
            "static": float(self.config["CHANGE_LANE"].get("dif_from_mid_multiplier_static")),
            "dynamic": float(self.config["CHANGE_LANE"].get("dif_from_mid_multiplier_dynamic")),
        }
        self.smoothing_iterations = self.max_smoothing_iterations
        self.smoothing_step = 0
        self.min_smoothing_factor = float(self.config["LANE_KEEPING"].get("min_smoothing_factor"))
        self.max_smoothing_factor = float(self.config["LANE_KEEPING"].get("max_smoothing_factor"))
        self.smoothing_factor = np.linspace(0.1, 1, self.smoothing_iterations)

        self.roundabout = False

        self.max_lane_changing_tries = {
            "static": int(self.config["CHANGE_LANE"].get("max_lane_changing_tries_static")),
            "dynamic": int(self.config["CHANGE_LANE"].get("max_lane_changing_tries_dynamic")),
        }

        # Calculate the survival function using the complementary cumulative distribution function (ccdf)
        cdf = norm.cdf(self.plot_y_norm, mu, sigma)
        sf = 1 - cdf
        self.sf = (sf - sf[-1]) / (sf[0] - sf[-1])

    # ------------------------------------------------------------------------------#

    def choose_405(self):
        self.bottom_perc = float(self.config["LANE_DETECT"].get("bottom_perc_405"))
        self.peaks_min_width = int(self.config["LANE_DETECT"].get("peaks_min_width_405"))
        self.peaks_max_width = int(self.config["LANE_DETECT"].get("peaks_max_width_405"))

        self.bottom_row_index = self.height - self.bottom_offset
        end = int((1 - self.bottom_perc) * self.height)
        self.step = int(-(self.height * self.bottom_perc / self.slices))
        self.real_slices = (end - self.bottom_row_index) // self.step
        self.top_row_index = self.bottom_row_index + self.real_slices * self.step

        self.plot_y = np.linspace(self.bottom_row_index, self.top_row_index, self.real_slices)
        self.plot_y_norm = np.linspace(0, 1, self.real_slices)

        self.get_error_weights = range(0, self.real_slices)

        mu = 0.5
        sigma = 0.4
        # Calculate the survival function using the complementary cumulative distribution function (ccdf)
        cdf = norm.cdf(self.plot_y_norm, mu, sigma)
        sf = 1 - cdf
        self.sf = (sf - sf[-1]) / (sf[0] - sf[-1])

        # For choosing lanes (Half image) (check if lanes difference is bigger/smaller than these params)
        self.bottom_width = int(self.config["LANE_KEEPING"].get("bottom_width_405"))
        self.top_width = int(self.config["LANE_KEEPING"].get("top_width_405"))

    def choose_455(self):
        self.bottom_perc = float(self.config["LANE_DETECT"].get("bottom_perc_455"))
        self.peaks_min_width = int(self.config["LANE_DETECT"].get("peaks_min_width_455"))
        self.peaks_max_width = int(self.config["LANE_DETECT"].get("peaks_max_width_455"))

        self.bottom_row_index = self.height - self.bottom_offset
        end = int((1 - self.bottom_perc) * self.height)
        self.step = int(-(self.height * self.bottom_perc / self.slices))
        self.real_slices = (end - self.bottom_row_index) // self.step
        self.top_row_index = self.bottom_row_index + self.real_slices * self.step

        self.plot_y = np.linspace(self.bottom_row_index, self.top_row_index, self.real_slices)
        self.plot_y_norm = np.linspace(0, 1, self.real_slices)
        self.get_error_weights = range(0, self.real_slices)

        mu = 0.5
        sigma = 0.4
        # Calculate the survival function using the complementary cumulative distribution function (ccdf)
        cdf = norm.cdf(self.plot_y_norm, mu, sigma)
        sf = 1 - cdf
        self.sf = (sf - sf[-1]) / (sf[0] - sf[-1])

        # For choosing lanes (Half image) (check if lanes difference is bigger/smaller than these params)
        self.bottom_width = int(self.config["LANE_KEEPING"].get("bottom_width_455"))
        self.top_width = int(self.config["LANE_KEEPING"].get("top_width_455"))

    # ------------------------------------------------------------------------------#

    def desired_lane(self, left_fit, right_fit, polynomial=2):
        """Caclulates multiple points on the desired lane that the car should follow.

        Parameters
        ----------
        left_fit : array
            The left polyfit
        right_fit : array
            The right polyfit

        Returns
        -------
        array
            The desired lane points
        """

        if left_fit is not None and right_fit is not None:
            # BOTH LANES (Desired lane is the middle lane of those lane)

            desire_lane = (
                (left_fit[0] + right_fit[0]) / 2 * self.plot_y**2
                + (left_fit[1] + right_fit[1]) / 2 * self.plot_y
                + (left_fit[2] + right_fit[2]) / 2
            )

        elif left_fit is None:
            # Create the desired lane depending only on the right lane , the bottom_width and top_width params
            desire_lane = np.ndarray([])

            top_width = self.top_width

            # right_fit = a,b,c where y = ax^2 +bx + c
            # if a < 0 => left turn => turn sharply => change top_width
            if right_fit[0] < -(self.max_coef_of_sharp_turn):
                top_width = int(top_width * self.sharp_turning_factor)
            elif right_fit[0] < -(self.min_coef_of_sharp_turn):
                add = (-right_fit[0] - self.min_coef_of_sharp_turn) / (
                    self.max_coef_of_sharp_turn - self.min_coef_of_sharp_turn
                )
                top_width = int(top_width * (1 + add * (self.sharp_turning_factor - 1)))

            cnt = 0
            for i in self.plot_y:

                fix = self.bottom_width - ((self.bottom_width - top_width) * self.plot_y_norm[cnt])
                cnt += 1

                # desire lane is on the left of the right_lane so we need to subtract the fix
                value = (right_fit[0] * i**2 + right_fit[1] * i + right_fit[2]) - fix

                desire_lane = np.append(desire_lane, value)

            desire_lane = desire_lane[1:]

        elif right_fit is None:
            # Create the desired lane depending only on the left lane , the bottom_width and top_width params
            desire_lane = np.ndarray([])

            top_width = self.top_width

            # left_fit = a,b,c where y = ax^2 +bx + c
            # if a > 0 => right turn => turn sharply => change top_width
            if left_fit[0] > self.max_coef_of_sharp_turn:
                top_width = int(top_width * self.sharp_turning_factor)
            elif left_fit[0] > self.min_coef_of_sharp_turn:
                add = (left_fit[0] - self.min_coef_of_sharp_turn) / (
                    self.max_coef_of_sharp_turn - self.min_coef_of_sharp_turn
                )
                top_width = int(top_width * (1 + add * (self.sharp_turning_factor - 1)))

            cnt = 0
            for i in self.plot_y:

                fix = self.bottom_width - ((self.bottom_width - top_width) * self.plot_y_norm[cnt])
                cnt += 1

                value = (left_fit[0] * i**2 + left_fit[1] * i + left_fit[2]) + fix

                desire_lane = np.append(desire_lane, value)

            desire_lane = desire_lane[1:]

        if (self.turn_left or self.turn_right) and self.change_lane_stage_1:
            smooth_lane = ((self.width // 2) - desire_lane) * (self.sf)
            desire_lane += smooth_lane
        elif self.change_lane_stage_2:
            overtake_type = self.var.get_overtake_type()
            dif_from_mid = (self.width // 2 - desire_lane[0]) * self.change_lane_mid[overtake_type]
            desire_lane += dif_from_mid

        return desire_lane.astype(np.int32)

    def visualize_desire_lane(self, frame, plot_x, bgr_colour=(50, 205, 50)):
        """Visualization of the desired lane that the car should be following.

        Parameters
        ----------
        frame : array
            input image
        plot_x : array
            having the x values of the points of the desired lane.
        bgr_colour : tuple
            An optional parameter representing the color of the desire lane.
        """

        if frame is not None:

            # buttom to top
            plot_y = self.plot_y.astype(np.int32)

            for i in range(1, len(plot_y) - 1):

                new = (int(plot_x[i]), int(plot_y[i]))
                prev = (int(plot_x[i - 1]), int(plot_y[i - 1]))

                cv2.line(frame, prev, new, bgr_colour, thickness=3)

    def get_error(self, desired):
        """Calculates the error between the setpoint (desired lane) and the measured lane position (the actual lane position of the vehicle)

        Parameters
        ----------
        desired : array
            The points of the desired lane

        Returns
        -------
        float
            calculated error
        """

        if len(desired) != 0:
            weighted_mean = np.average(desired, weights=[*self.get_error_weights])

            # The actual lane is the middle lane.
            center = int(self.width / 2.0)
            error = weighted_mean - center

            return error
        return 0

    # ------------------------------------------------------------------------------#
    def change_lane_maneuver(self, lanes_detection, direction, smoothing_iterations=None):

        destination = {"right": 1, "left": -1}

        if smoothing_iterations and smoothing_iterations < self.max_smoothing_iterations:
            self.smoothing_iterations = smoothing_iterations
            self.smoothing_factor = np.linspace(
                self.min_smoothing_factor, self.max_smoothing_factor, self.smoothing_iterations
            )

        self.angle, frame = self.change_lane(lanes_detection, destination=destination[direction])

        text_color = (0, 0, 139) if direction == "left" else (139, 0, 0)

        cv2.putText(
            frame,
            f"Changing to {direction} lane...",
            (int(0.15 * self.width), int(0.1 * self.height)),
            cv2.FONT_HERSHEY_COMPLEX,
            1,
            text_color,
            1,
        )

        return self.angle, frame

    def change_lane(self, lanes_detection, destination):

        threshold = self.change_lane_threshold
        count = self.change_lane_count

        # extract data from lane detection
        frame = lanes_detection["frame"]
        left_lane = lanes_detection["left"]
        right_lane = lanes_detection["right"]
        left_coef = lanes_detection["left_coef"]
        right_coef = lanes_detection["right_coef"]

        # Choose turning left/right
        self.turn_left = True if (destination == -1 and not self.turn_right) else False
        self.turn_right = True if (destination == 1 and not self.turn_left) else False

        # If both are False => We still are on the middle of changing lanes so act accordingly
        if not self.turn_left and not self.turn_right:
            # if stage_1 => reset params and exit
            if self.change_lane_stage_1:
                self.reset_change_lane_params()
                self.angle, frame = self.lane_keeping(lanes_detection)

                self.log.debug(
                    ">>>> Lane change interrupted - remaining in current lane due to conflicting turn signal.\n"
                )
                return self.angle, frame

            # else (stage_2) => reset params and run change_lane depending on the last call
            else:
                self.reset_change_lane_params()
                self.turn_left = True if destination == -1 else False
                self.turn_right = True if destination == 1 else False

        # Find mean of each lane
        # maybe find the middle point of the polyfits (from coefs)
        mean_right = 0 if not right_lane else float(np.average(right_lane))
        mean_left = 0 if not left_lane else float(np.average(left_lane))

        # if we have not change lanes (From the beginning)
        if self.change_lane_stage_1:

            # For left turn
            if self.turn_left:
                # Because we want to turn left check if left lane became right
                left_lane_changed_to_right = (
                    self.prev_mean_left + threshold > mean_right
                    and self.prev_mean_left > 0
                    and mean_right > 0
                )

                if left_lane_changed_to_right:
                    self.change_lane_stage_1 = False
                    self.change_lane_stage_2 = True
                    # else stay in stage 1

            # for right turn
            elif self.turn_right:

                # check if right lane became the left one
                right_lane_changed_to_left = (
                    self.prev_mean_right - threshold < mean_left
                    and self.prev_mean_right > 0
                    and mean_left > 0
                )

                if right_lane_changed_to_left:
                    self.change_lane_stage_1 = False
                    self.change_lane_stage_2 = True

        # if lanes changed again as they were before
        elif self.change_lane_stage_2:

            if self.turn_left:

                # if right lane changes to left
                right_lane_changed_to_left = (
                    self.prev_mean_right - threshold < mean_left
                    and self.prev_mean_right > 0
                    and mean_left > 0
                )

                if right_lane_changed_to_left:
                    self.change_lane_stage_2 = False
                    self.change_lane_stage_1 = True
                    # else stay in stage 2

            elif self.turn_right:

                # if left lane changes to right
                left_lane_changed_to_right = (
                    self.prev_mean_left + threshold > mean_right
                    and self.prev_mean_left > 0
                    and mean_right > 0
                )

                if left_lane_changed_to_right:
                    self.change_lane_stage_2 = False
                    self.change_lane_stage_1 = True

        # [Stage 1] When left/right lanes are still the same
        if self.change_lane_stage_1:

            # Check if with the detected lanes we can switch lanes
            if (self.turn_left and not left_lane) or (self.turn_right and not right_lane):
                # initialize params for next run and exit

                self.reset_change_lane_params()
                self.angle, frame = self.lane_keeping(lanes_detection)
                self.var.set_lane_change_completed(True)
                self.log.debug(
                    "\n\n\n\n>>>> Unable to change lanes - correct lane did not detected.\n\n\n\n"
                )

                return self.angle, frame

            if self.turn_left:
                # Make left --> None, right lane --> left and then do lane keeping
                self.angle, error, frame = self.pid_controller(None, left_coef, frame)
                self.count = 0

            elif self.turn_right:
                # Make left lane --> right, right --> None and then do lane keeping
                self.angle, error, frame = self.pid_controller(right_coef, None, frame)
                self.count = 0

        # [Stage 2] When lanes have changed only once (Do lane keeping without changing anything, wait for some frames and then exit)
        elif self.change_lane_stage_2:

            self.angle, error, frame = self.pid_controller(left_coef, right_coef, frame)

            # if we are on stage 2 for the required frames (change params for next run and exit)
            if self.count >= count:
                self.reset_change_lane_params()
                self.var.set_lane_change_completed(True)
                self.log.debug(">>>> Lane change completed successfully.\n")
            else:
                self.count += 1

        self.fix_angle()
        self.last_angle = self.angle

        self.prev_mean_left = mean_left
        self.prev_mean_right = mean_right

        return self.angle, frame

    def reset_change_lane_params(self):
        self.turn_left = False
        self.turn_right = False
        self.count = 0
        self.change_lane_stage_1 = True
        self.change_lane_stage_2 = False
        self.prev_mean_left = 0
        self.prev_mean_right = 0
        self.smoothing_iterations = self.max_smoothing_iterations
        self.smoothing_step = 0
        self.smoothing_factor = np.linspace(
            self.min_smoothing_factor,
            self.max_smoothing_factor,
            self.smoothing_iterations,
        )

    # ------------------------------------------------------------------------------#

    def pid_controller(self, left, right, frame=None):
        """Performs the angle calculation based on the left and right polyfits. Steps :
        1) create a polyfit for the desired lane that the car should follow.
        2) extract the points of that lane and the lane that the car currently follows.
        3) find the error of those two points (width diffrence in the same height)
        4) adjust steering angle based on the error and PID controller's parameters

        Parameters
        ----------
        left : array
            Left polyfit
        right : array
            Right polyfit
        frame : array
            The frame with the lines

        Returns
        -------
        float
            calculated angle
        float
            calculated error
        array
            The frame with the desire lane
        """

        if left is None and right is None:
            angle = self.last_angle
            error = 0
            self.log.warning("No lanes found")
        else:
            desired_lane = self.desired_lane(left, right)

            if self.print_desire_lane:
                self.visualize_desire_lane(frame, desired_lane)

            error = self.get_error(desired_lane)
            angle = 90 - math.degrees(math.atan2(self.height, error))

            # If we change lanes --> Smooth the lane
            if (self.turn_left or self.turn_right) and self.change_lane_stage_1:
                if self.smoothing_step < self.smoothing_iterations:
                    angle = angle * self.smoothing_factor[self.smoothing_step]
                    self.smoothing_step += 1

        self.last_angle = angle
        return angle, error, frame

    def fix_angle(self):
        """Fixes the angle so the car changes its position smoothly.
        - simple PD control on the angle (not activated now)
        - rolling average on the last N calculated angles (activated)
        """

        self.steer_value_list.insert(0, self.angle)
        if len(self.steer_value_list) == self.median_constant:
            self.steer_value_list.pop()

        median = np.average(self.steer_value_list)
        self.angle = median
        self.last_angle = self.angle

    def lane_keeping(self, lanes_detection):
        """Performs the pipeline for the lanekeeping.
        1) Perform Lane Detection
        2) Calculate the steering angle for lane keeping
        3) Smooth the angle

        Parameteres
        -----------
        lanes_detection : dictionary
            Result of lane detection : frame, left_coef, right_coef

        Returns
        -------
        float
            calculated steering angle
        array
            lane keeping output image
        """

        frame = lanes_detection["frame"]
        left_coef = lanes_detection["left_coef"]
        right_coef = lanes_detection["right_coef"]

        if self.turn_left:
            self.angle, frame = self.change_lane_maneuver(lanes_detection, direction="left")

        elif self.turn_right:
            self.angle, frame = self.change_lane_maneuver(lanes_detection, direction="right")

        else:

            if self.var.get_intersection_started() and not self.var.get_intersection_takeover():
                # if left_coef is not None and right_coef is not None:
                if self.var.get_post_roadblock_lane() == "left" or self.var.get_bamby_started():
                    left_coef = None
                else:
                    right_coef = None
            else:
                trust_left = lanes_detection["trust_left"]
                trust_right = lanes_detection["trust_right"]

                left_coef = left_coef if trust_left else None
                right_coef = right_coef if trust_right else None

            self.angle, error, frame = self.pid_controller(left_coef, right_coef, frame)

            if not self.speedrun:
                self.fix_angle()
            self.last_angle = self.angle

        self.angle = max(min(self.max_lk_steer, self.angle), -self.max_lk_steer)

        return self.angle, frame

File: src/LaneKeeping/lanekeeping.cpp
//
// Created by giannishorgos on 04/01/2024.
//
#include <opencv2/opencv.hpp>
#include <stdio.h>
// #include "../laneDetection/helpers.hpp"
#include "lanekeeping.hpp"
#include <iostream>
#include <vector>
#include <eigen/Eigen/Dense>


laneKeeping::laneKeeping(int width, int height, std::string &camera){
    std::cout << ">>> LaneKeeping Constructor called..." << std::endl;

    this->img_width = width;
    this->img_height = height;
    this->camera = camera;

    // LK
    this->median_constant = 2;
    this->print_desire_lane = false;
    this->bottom_width_455 = 190;
    this->bottom_width_405 = 145;
    this->top_width_455 = 85;
    this->top_width_455 = 80;
    this->max_coef_of_sharp_turn = 0.01;
    this->min_coef_of_sharp_turn = 0.0001;
    this->sharp_turning_factor = 2;
    this->mu = 0.5;
    this->sigma = 0.4;
    this->min_smoothing_factor = 0.2;
    this->max_smoothing_factor = 0.8;

    // PARAMS
    this->max_lk_steer = 23;

    
    // LD 
    this->slices = 20;
    this->bottom_offset = 10;
    this->bottom_perc_455 = 0.5;
    this->bottom_perc_405 = 0.45;
    this->peaks_min_width_455 = 3;
    this->peaks_max_width_405 = 2;
    this->peaks_max_width_455 = 20;
    this->peaks_max_width_405 = 15;

    // HELP
    this->already_initialized = false;
    this->angle = 0.0;
    this->last_angle = 0.;
   
    this->left_lane = new Polynomial;
    this->right_lane = new Polynomial;
    this->desire_lane_ptr = nullptr;
    
    if (this->camera == "405")
        laneKeeping::choose_405();
    else if (this->camera == "455")
        laneKeeping::choose_455();
    else
        throw std::runtime_error(">>> LANEKEEPING ERROR on Initialization :\n\t Can't choose a camera. No correct camera selected!");

    // std::cout << "Real Slices: " << this->real_slices << ", Slices: " << this->slices << std::endl;
}

void laneKeeping::choose_405(){
    std::cout << "405 camera selected successfully" << std::endl;

    this->bottom_perc = this->bottom_perc_405;
    this->peaks_min_width = this->peaks_min_width_405;
    this->peaks_max_width = this->peaks_max_width_405;

    this->bottom_row_index = this->img_height - this->bottom_offset;
    int end = int(float(this->img_height) * (1 - this->bottom_perc));
    this->step = int(-(float(this->img_height) * this->bottom_perc / float(this->slices)));
    this->real_slices = int((end - this->bottom_row_index) / this->step);
    this->top_row_index = this->bottom_row_index + this->real_slices * this->step;

    if (this->already_initialized){
        delete[] this->plot_y;
        delete[] this->plot_y_norm;
        delete[] this->desire_lane;
    }
    // this->plot_y = linspace(this->bottom_row_index, this->top_row_index, this->real_slices + 1);
    this->plot_y = linspace(this->bottom_row_index, this->top_row_index, this->real_slices);
    this->plot_y_norm = linspace(0,1,this->real_slices);
    this->desire_lane = new int[this->real_slices];

    this->bottom_width = this->bottom_width_405;
    this->top_width = this->top_width_405;

    this->already_initialized = true;
}

void laneKeeping::choose_455(){
    std::cout << "455 camera selected successfully" << std::endl;

    this->bottom_perc = this->bottom_perc_455;
    this->peaks_min_width = this->peaks_min_width_455;
    this->peaks_max_width = this->peaks_max_width_455;

    this->bottom_row_index = this->img_height - this->bottom_offset;
    int end = int(float(this->img_height) * (1 - this->bottom_perc));
    this->step = int(-(float(this->img_height) * this->bottom_perc / float(this->slices)));
    this->real_slices = int((end - this->bottom_row_index) / this->step);
    this->top_row_index = this->bottom_row_index + this->real_slices * this->step;

    if (this->already_initialized){
        delete[] this->plot_y;
        delete[] this->plot_y_norm;
        delete[] this->desire_lane;
    }
    // this->plot_y = linspace(this->bottom_row_index, this->top_row_index, this->real_slices + 1);
    this->plot_y = linspace(this->bottom_row_index, this->top_row_index, this->real_slices);
    this->plot_y_norm = linspace(0, 1, this->real_slices);
    this->desire_lane = new int[this->real_slices];

    this->bottom_width = this->bottom_width_455;
    this->top_width = this->top_width_455;

    this->already_initialized = true;
}

laneKeeping::~laneKeeping(){
    std::cout << ">>> LK Destructor called..." << std::endl;
    delete [] this->plot_y;
    delete[] this->plot_y_norm;
    delete[] this->desire_lane;
}

double laneKeeping::lane_keeping(LaneDetectionResults results){

    // results.print_lane_detection_result();
    cv::Mat frame = results.frame;
    this->left_lane->copyPolynomial(results.left_poly);
    this->right_lane->copyPolynomial(results.right_poly);

    // Set lanes as accepted if they are trusted
    this->left_lane->accepted = results.trust_left;
    this->right_lane->accepted = results.trust_right;
        
    this->angle = laneKeeping::pid_controller(this->left_lane, this->right_lane, frame);
    this->last_angle = this->angle;

    // std::cout << ">>> Angle: " << this->angle <<std::endl;
    return this->angle;
}

double laneKeeping::pid_controller(Polynomial* left, Polynomial* right, cv::Mat &frame){
    double angle;

    if(not left->accepted and not right->accepted){
        angle = this->last_angle;
        std::cout<< ">>> (LK WARNING!!!) NO LANES FOUND"<< std::endl;
    }else{
        this->desire_lane_ptr = laneKeeping::calculate_desire_lane(left, right);
        if(this->print_desire_lane){
            laneKeeping::visualize_desire_lane(frame, this->desire_lane);
        }
        
        double error = laneKeeping::get_error(this->desire_lane_ptr);
        
        angle = 90 - (180 / M_PI) * atan2(this->img_height, error);

    }
    return angle;
}

int* laneKeeping::calculate_desire_lane(Polynomial* left, Polynomial* right){
    if(left->accepted and right->accepted){
        // BOTH LANES FOUND
        for(int i=0; i< this->real_slices; i++){
            this->desire_lane[i] = (
             (left->a + right->a)/2 * this->plot_y[i] * this->plot_y[i] +
             (left->b + right->b)/2 * this->plot_y[i] + 
             (left->c + right->c)/2
             );
        }
    }else if(not left->accepted){
        // Create the desired lane depending only on the right lane , the bottom_width and top_width params
        
        int top_width = this->top_width;
        int fix;

        if(right->a < -(this->max_smoothing_factor)){
            top_width = int(top_width * this->sharp_turning_factor);
        }else if(right->a < -(this->min_coef_of_sharp_turn)){
                int add = (-right->a - this->min_coef_of_sharp_turn) / (
                        this->max_coef_of_sharp_turn - this->min_coef_of_sharp_turn);
                top_width = int(top_width * (1 + add * (this->sharp_turning_factor - 1)));
        }

        for(int i=0; i< this->real_slices; i++){
            fix = this->bottom_width - ((this->bottom_width - top_width)* this->plot_y_norm[i]);
            this->desire_lane[i] = right->evaluate(plot_y[i]) - fix;
        }

    }else{
        int top_width = this->top_width;
        int fix;

        if(left->a > this->max_smoothing_factor){
            top_width = int(top_width * this->sharp_turning_factor);
        }else if(left->a > this->min_coef_of_sharp_turn){
            int add = (left->a - this->min_coef_of_sharp_turn) / (
                    this->max_coef_of_sharp_turn - this->min_coef_of_sharp_turn);
            top_width = int(top_width * (1 + add * (this->sharp_turning_factor - 1)));
        }


        for(int i=0; i< this->real_slices; i++){            
            fix = this->bottom_width - ((this->bottom_width - top_width)* this->plot_y_norm[i]);
            this->desire_lane[i] = left->evaluate(plot_y[i]) + fix;
        }
    }

    return this->desire_lane;
}

double laneKeeping::get_error(int* desired){
    if(desired!=nullptr){
        double mean = get_mean(desired);

        // The actual lane is the middle lane
        int center = this->img_width/2;

        double error = mean - center;
        return error;
    }
    return 0;
}

double laneKeeping::get_mean(int* desire){
    double sum=0;
    for(int i=0; i< this->real_slices; i++){
        sum += desire[i];
    }
    return sum/this->real_slices;
}

void laneKeeping::visualize_desire_lane(cv::Mat &frame, int* plot_x){
    if(this->desire_lane_ptr == nullptr)
        return;

    cv::Point point_1, point_2;
    cv::Scalar colour = cv::Scalar(50, 205 ,50);

    for(int i=1; i<this->real_slices; i++){
        point_1.x = plot_x[i];
        point_1.y = plot_y[i];
        point_2.x = plot_x[i-1];
        point_2.y = plot_y[i-1];

        cv::line(frame, point_1, point_2, colour, 3);
    }
    cv::Point p1(int(this->img_width/2), this->img_height);
    point_1.x = int(this->img_width/2);
    cv::line(frame, p1, point_1, cv::Scalar(50,50,205), 3);
}

double* laneKeeping::linspace(double start, double end, int numPoints) {
    if (numPoints <= 1) {
        double* result = new double[1];
        result[0] = start;
        return result;
    }
    
    double* result = new double[numPoints];
    double step = (end - start) / (numPoints - 1);
    for (int i = 0; i < numPoints; ++i) {
        result[i] = start + i * step;
    }
    
    return result;
}

double* laneKeeping::copy_coeffs(double* coef_dest, double* coef){
    /**
     * Coef ptr = null if coef == null
     * else coef_data
    */
    if(coef != nullptr){
        coef_dest[0] = coef[0];
        coef_dest[1] = coef[1];
        coef_dest[2] = coef[2];
        return coef_dest;
    }else{
        return nullptr;
    }
}
